### PR TITLE
apk: cumulative fixes

### PR DESCRIPTION
**Pull request by Ansuel**
Cumulative fixes for APK implementation.

---

@aparcar about sign we need to understand how to handle this... I'm still convinced the real and correct solution for this would be to have some kind of additional app in APK to append the signature to the pkg later... With this implemented we can just sign every package when we build index and solve this race condition + drop all the --allow-unstrusted stuff.

### CODE REVIEW

**Comment by dangowrt** [comment]
Commit "package: skip sign of packages for APK" can be dropped now as keypair should always be present after [this fix](https://github.com/openwrt/openwrt/commit/9a11bc3682dd7e95eb0534da9e2dbe3cdaa1c6de).

Commit "apk: fix problem with scripts.tar using base64 CSUM" got a typo in the commit description: "manaully" -> "manually"

Commit "include/rootfs: rework handling of post-install scripts for APK" removes `scripts.tar` which may not only contain post-install, but also other types of scripts which are relevant when the package is uninstalled or being upgraded. Hence we should only get rid of the post-install and pre-install scripts in `scripts.tar` and maybe have APK handle that instead of taking care of it during rootfs image creation.

All the rest looks good to me.

**Comment by aparcar** [comment]
Hi, thanks for the fixes. I'm unsure this can be fixed like this once we let the Buildbot do the work. Let me explain: Right now the buildbot workers use temporary keys which are never used on the device. Instead the package index is transferred the build buildbot master, signed, copied back and then transferred to our download server.

The same approach should work with APK, that it signs an index without modifying the included packages. Problem with APK right now is that it signs every single package (plus index). With that approach we'd need to copy all our packages to the buildbot master etc.

I suggest we patch APK that it allows unsigned packages without a warning, relying entirely on the signed package index. This, with he assumption that the index holds checksums of the package itself.

**Comment by dangowrt** [comment]
@aparcar The idea behind https://github.com/openwrt/openwrt/commit/9a11bc3682dd7e95eb0534da9e2dbe3cdaa1c6de is that the private key will only be generated by the build system if it is missing, and the public key will only be generated also only if missing and also if the timestamp of the private key is newer than the one of the existing public key.
In this way we can integrate buildbot generating packages as well as image with a PKI -- sadly, as I noticed now, APK doesn't use X.509 but only a simple EC public key, hence key delegation is (just like with opkg) *not* possible.

**Comment by dangowrt** [comment]
> I suggest we patch APK that it allows unsigned packages without a warning, relying entirely on the signed package index. This, with he assumption that the index holds checksums of the package itself.

I think this is not a good idea, it will reduce the security level of APK which we would gain otherwise.

File: include/rootfs.mk [hunk 1]
```diff
@@ -115,6 +121,8 @@ define prepare_rootfs
 	rm -rf \
 		$(1)/boot \
 		$(1)/tmp/* \
+		$(1)/lib/apk/db/scripts.tar \
```
Comment (dangowrt) [inline include/rootfs.mk:124]: We will still need `scripts.tar` for all the other scripts, to be run on removal or upgrade of the package.

**Comment by Ansuel** [comment]
> Commit "include/rootfs: rework handling of post-install scripts for APK" removes `scripts.tar` which may not only contain post-install, but also other types of scripts which are relevant when the package is uninstalled or being upgraded. Hence we should only get rid of the post-install and pre-install scripts in `scripts.tar` and maybe have APK handle that instead of taking care of it during rootfs image creation.

@dangowrt looking at how we are currently building the packet we don't handle other kind of script i think... Have to check the wiki for what we support with opkg. https://github.com/openwrt/openwrt/blob/main/include/package-pack.mk#L319

Are we sure script.sh should also contain those? From what I was reading in the code scripts.tar is just a container that APK consume for any kind of pending script. But i'm probably wrong on that.

File: include/rootfs.mk (hunk 1 above)
Comment (Ansuel) [reply include/rootfs.mk:124]: I will just edit the tar and remove all the relevant script but I opened the scripts.tar and there were only post-install and pre-deinstall.

Ok i guess pre-deinstall should not be removed...

**Comment by aparcar** [comment]
Please elaborate how you'd like to implement this.

> @aparcar The idea behind [9a11bc3](https://github.com/openwrt/openwrt/commit/9a11bc3682dd7e95eb0534da9e2dbe3cdaa1c6de) is that the private key will only be generated by the build system if it is missing, and the public key will only be generated also only if missing and also if the timestamp of the private key is newer than the one of the existing public key. 

I understand the commit and think it's a cleaner solution compared to hiding it in base-files.

> In this way we can integrate buildbot generating packages as well as image with a PKI -- sadly, as I noticed now, APK doesn't use X.509 but only a simple EC public key, hence key delegation is (just like with opkg) _not_ possible.

I haven't looked into the details but to my knowledge APK also support other kinds of keys, whatever OpenSSL/MbedTLS works with, Alpine is using some RSA keys if I'm not mistaken.

> I think this is not a good idea, it will reduce the security level of APK which we would gain otherwise.

I'm not so sure about this. What implementation do you suggest and how does it makes things more secure.

From my perspective the buildbot master or whatever dedicated device we use for signing got a reduced attack surface, talking to @ynezz we may even use hardware keys to secure the private keys.

With your suggestion we'd have to come up with a PKI where the issue of revoking is entirely unsolved within OpenWrt. If that would have been a focus we could have also implemented it with OPKG, `usign` and `ucert`. It's using `--verify-program` after all.

So from my point of view a signed index containing the SHA512 sums of the packages has the same security level as additionally signed packages for default usage (i.e. using feeds). Special use-cases like `apk add https://example.com/package.apk` wouldn't profit from this, but if desired those packages could be signed by the creator individually anyway.

Overall, I'm trying to keep the changes to integrate "APK" small'ish, there was already plenty of outcry and possible breakage for the enforced version schema. If we now rework the entire buildbot setup we might as-well start using the incremental approach used by Alpine anyway.

File: package/Makefile [hunk 2]
```diff
@@ -81,7 +81,7 @@ endif
 ifndef SDK
   $(curdir)//compile = $(STAGING_DIR)/.prepared $(BIN_DIR)
 ifneq ($(CONFIG_USE_APK),)
-  $(curdir)/compile: $(curdir)/system/apk/host/compile $(BUILD_KEY_APK_SEC) $(BUILD_KEY_APK_PUB)
```
Comment (ynezz) [inline package/Makefile:84]: BTW I've hit that as well and come up with the following solution (untested fully yet) https://github.com/ynezz/openwrt/commit/ff6d450321af205de6820b5b4ac2b392b9cdecc4

Comment (ynezz) [reply package/Makefile:84]: Lets see how it goes https://buildbot.staging.openwrt.org/images/#/builders/3/builds/2

Comment (Ansuel) [reply package/Makefile:84]: @ynezz did you verified that this works by removing the hostpkg apk build_dir and apk tool from staging_dir?

$(curdir)//compile is actually different to $(curdir)/compile ?

Comment (dangowrt) [reply package/Makefile:84]: @Ansuel It's a hack to have two different Make targets for the same file...

Comment (dangowrt) [reply package/Makefile:84]: @ynezz Looks good now, doesn't it? In production we will want to provision the key instead of generating a new one for every single build. Or find a way to skip signing at the pkgbuild stage and have another instance not running any 3rd-party code do that part separately -- at the cost of having to upload all packages to that other instance...

Comment (Ansuel) [reply package/Makefile:84]: if we really want to to sign each package... maybe there the tool adbsign with --reset-signature that we maybe use to sign .apk later when they are packaged... but i still have to try that...

Currently busy with damn apk not being able to handle long filename in the script.tar that is a big red flag...

Example:
image

Comment (ynezz) [reply package/Makefile:84]: > Looks good now, doesn't it?

Yep, at least it builds now.

> Or find a way to skip signing at the pkgbuild stage and have another instance not running any 3rd-party code do that part separately

Keys on disk = not your keys anymore :-) Thus I've started [Using Nitrokey 3A Mini for build artifact signing key storage](https://lists.openwrt.org/pipermail/openwrt-devel/2023-May/041044.html).

**Comment by aparcar** [comment]
@Ansuel this branch here will be used for buildbot.staging.openwrt.org to and uploaded to downloads.staging.openwrt.org. Please don't close it but let's work together on this branch until APK is fully working.

**Comment by aparcar** [comment]
Actually the buildbot will fetch your master branch, please push those changes over there. Sorry for the extra loop.

**Comment by Ansuel** [comment]
BTW for everyone interested, things are going well. Getting very close on catching all the APK corner case. Images already works.

The only big task to do is sign and verify.

File: package/network/services/dropbear/Makefile [hunk 3]
```diff
@@ -224,7 +224,6 @@ define Package/dropbear/install
 	$(INSTALL_CONF) ./files/dropbear.config $(1)/etc/config/dropbear
 	$(INSTALL_DIR) $(1)/etc/init.d
 	$(INSTALL_BIN) ./files/dropbear.init $(1)/etc/init.d/dropbear
-	$(INSTALL_DIR) $(1)/usr/lib/opkg/info
```
Comment (KanjiMonster) [inline package/network/services/dropbear/Makefile:227]: If you want to point to a revision for a Fixes tag, it's this one: a377aa9ab534ed8fbbc6bbd867bfd66566da73ce. It clearly forgot to remove this line (but nobody noticed until now)

File: include/package-pack.mk [hunk 4]
```diff
@@ -329,6 +329,10 @@ else
 		rm -rf $$(IDIR_$(1))/CONTROL/conffiles; \
 	fi
 
+    ifneq ($$(CONFIG_IPK_FILES_CHECKSUMS),)
+	if [ -f $$(IDIR_$(1))/CONTROL/files-sha256sum ]; then mv -f $$(IDIR_$(1))/CONTROL/files-sha256sum $$(IDIR_$(1))/lib/apk/packages/$(1).files-sha256sum; fi
```
Comment (dangowrt) [inline include/package-pack.mk:333]: Isn't that effectively the same thing already provided as built-in feature of APK?
Afaik `apk audit` is exactly for that and we don't need additional files with hashes.

Comment (aparcar) [reply include/package-pack.mk:333]: Yea I think you can drop this commit entirely... sorry.

Comment (Ansuel) [reply include/package-pack.mk:333]: Ok I will change this commit to depend on !USE_APK

File: package/system/apk/patches/0010-database-always-use-CSUM-HEX-for-scriptdb-tar-entry.patch [hunk 5]
```diff
@@ -0,0 +1,75 @@
+From 64cebdc05c07e890f85daba2c89d6a6d9a936e2e Mon Sep 17 00:00:00 2001
+From: Christian Marangi <ansuelsmth@gmail.com>
+Date: Wed, 22 May 2024 16:41:13 +0200
+Subject: [PATCH] database: always use CSUM HEX for scriptdb tar entry
```
Comment (aparcar) [inline package/system/apk/patches/0010-database-always-use-CSUM-HEX-for-scriptdb-tar-entry.patch:4]: This was merged upstream, please drop the patch.

File: include/rootfs.mk [hunk 6]
```diff
@@ -80,9 +80,16 @@ define prepare_rootfs
 	@( \
 		cd $(1); \
 		if [ -n "$(CONFIG_USE_APK)" ]; then \
-		$(STAGING_DIR_HOST)/bin/tar -xf ./lib/apk/db/scripts.tar --wildcards "*.post-install" -O > script.sh; \
-		chmod +x script.sh; \
-		IPKG_INSTROOT=$(1) $$(command -v bash) script.sh; \
+		$(STAGING_DIR_HOST)/bin/tar -C ./lib/apk/db/ -xf ./lib/apk/db/scripts.tar --wildcards "*.post-install"; \
+		for script in ./lib/apk/db/*.post-install; do \
```
Comment (aparcar) [inline include/rootfs.mk]: If we use the same mechanism anyway, can't we use the same for loop, too?

File: include/package-pack.mk [hunk 7]
```diff
@@ -205,6 +205,8 @@ $(_endef)
 	rm -rf $$(IDIR_$(1))
 ifeq ($$(CONFIG_USE_APK),)
 	$$(call remove_ipkg_files,$(1),$$(call opkg_package_files,$(call gen_package_wildcard,$(1))))
+else
+	$$(call remove_ipkg_files,$(1),$$(call apk_package_files,$(call gen_package_wildcard,$(1))))
```
Comment (aparcar) [inline include/package-pack.mk:209]: This is fine with me, however we could keep in mind that APK can actually handle multiple versions of the same package and eventually support that.

File: include/package-pack.mk [hunk 8]
```diff
@@ -309,6 +309,20 @@ else
...
+	# Append stuff from it and delete the CONTROL directory since everything else should be migrated
+	if [ -f $$(IDIR_$(1))/CONTROL/conffiles ]; then \
+		cat $$(IDIR_$(1))/CONTROL/conffiles >> $$(IDIR_$(1))/lib/apk/packages/$(1).conffiles; \
+		rm -rf $$(IDIR_$(1))/CONTROL/conffiles; \
+	fi
+
+	if [ -z "$$$$(ls -A $$(IDIR_$(1))/CONTROL 2>/dev/null)" ]; then \
+		rm -rf $$(IDIR_$(1))/CONTROL; \
+	else \
+		echo "CONTROL directory $$(IDIR_$(1))/CONTROL is not empty! This is not right and should be checked!" >&2; \
```
Comment (aparcar) [inline include/package-pack.mk:339]: This is cute but do we actually see those anywhere? I'd just go for `rm -rf CONTROL`

File: tools/tar/patches/001-Avoid-excess-lseek-etc.patch [hunk 9]
```diff
@@ -0,0 +1,491 @@
+From 66be5a789e70f911170017754fc51e499998f90e Mon Sep 17 00:00:00 2001
+From: Paul Eggert <eggert@cs.ucla.edu>
+Date: Sun, 14 Aug 2022 16:32:26 -0700
+Subject: [PATCH] Avoid excess lseek etc.
```
Comment (aparcar) [inline tools/tar/patches/001-Avoid-excess-lseek-etc.patch:4]: Any chance this flies upstream?

File: package/system/apk/patches/0020-tar-generalize-apk_tar_write_padding-size-handling.patch [hunk 10]
```diff
@@ -0,0 +1,65 @@
+From 7ebc1eb6de2a864cb7ec80cc15c082320b305890 Mon Sep 17 00:00:00 2001
+From: Christian Marangi <ansuelsmth@gmail.com>
+Date: Thu, 23 May 2024 19:09:27 +0200
+Subject: [PATCH 1/4] tar: generalize apk_tar_write_padding size handling
```
Comment (aparcar) [inline package/system/apk/patches/0020-tar-generalize-apk_tar_write_padding-size-handling.patch:4]: This was merged upstream, please drop the patch.

File: package/kernel/linux/modules/hwmon.mk [hunk 11]
```diff
@@ -198,7 +198,7 @@ define KernelPackage/hwmon-g762
 endef
 
 define KernelPackage/hwmon-g762/description
- Kernel module for Global Mixed-mode Technology Inc G762 and G763 fan speed PWM controller chips.
+ Kernel module for Global Mixed-mode Technology Inc G761/G762/G763 fan speed PWM controller chips.
```
Comment (aparcar) [inline package/kernel/linux/modules/hwmon.mk:201]: This patch seems unrelated to APK, please drop it from this series.

File: include/package-pack.mk (hunk 8 above)
Comment (Ansuel) [reply include/package-pack.mk:339]: I would like to keep it cause I have seen lots of strange thing done by package... It's a good way to catch them.

File: include/package-pack.mk (hunk 7 above)
Comment (Ansuel) [reply include/package-pack.mk:209]: Can this be a problem for stable release where we update some package not marked as hold?

File: config/Config-build.in [hunk 12]
```diff
@@ -69,6 +69,7 @@ menu "Global build settings"
 		default y
 
 	config USE_APK
+		select PACKAGE_apk-mbedtls
```
Comment (dangowrt) [inline config/Config-build.in:72]: Maybe `imply` is better so users can chose `apk-openssl` instead.

**Comment by ynezz** [comment]
BTW what do you think about following [apk: host: make the help available](https://github.com/ynezz/openwrt/commit/0af42c3c9d6080c37ca63da3c14e5bc86f2977a1) ? 

```
Currently, the apk utility lacks accessible help documentation, making
it cumbersome for human users:

 apk-tools 3.0.0_pre20240519, compiled for x86_64.

 ERROR: This apk-tools has been built without help

This absence of help forces users to delve into the apk's build
directory to understand its functionality. To enhance usability, we will
enable the help feature for the host build. The host environment can
accommodate the 3% increase in binary size for the added convenience.

On Ubuntu 22.04, x86_64 platform, the apk size increases by 17,816 bytes
(from 594,144 to 611,960 bytes), a 2.99% increase. This is a reasonable
trade-off for improved ergonomics.
```

**Comment by aparcar** [comment]
> BTW what do you think about following [apk: host: make the help available](https://github.com/ynezz/openwrt/commit/0af42c3c9d6080c37ca63da3c14e5bc86f2977a1) ? 
> 
> 
> 
> ```
> 
> Currently, the apk utility lacks accessible help documentation, making
> 
> it cumbersome for human users:
> 
> 
> 
>  apk-tools 3.0.0_pre20240519, compiled for x86_64.
> 
> 
> 
>  ERROR: This apk-tools has been built without help
> 
> 
> 
> This absence of help forces users to delve into the apk's build
> 
> directory to understand its functionality. To enhance usability, we will
> 
> enable the help feature for the host build. The host environment can
> 
> accommodate the 3% increase in binary size for the added convenience.
> 
> 
> 
> On Ubuntu 22.04, x86_64 platform, the apk size increases by 17,816 bytes
> 
> (from 594,144 to 611,960 bytes), a 2.99% increase. This is a reasonable
> 
> trade-off for improved ergonomics.
> 
> ```

I disabled it as a quick solution for failing builds due to concurrent dependency issues. I'll dive in deeper how to correctly depend on Lua being compiled before build APK

**Comment by ynezz** [comment]
> I disabled it as a quick solution for failing builds due to concurrent dependency issues. I'll dive in deeper how to correctly depend on Lua being compiled before build APK

Perhaps https://github.com/ynezz/openwrt/commit/c5037dca57#diff-35d31b1ae4d877b18ed95633f1cf31ae2027859662b254975533820ed213a2c4R20 ?

File: include/package-pack.mk (hunk 7 above)
Comment (aparcar) [reply include/package-pack.mk:209]: I guess, let's stick with what we know for now and please ignore my previous comment.

File: include/package-pack.mk (hunk 8 above)
Comment (aparcar) [reply include/package-pack.mk:330]: Sure why not, please proceed.

**Review by aparcar — APPROVED** [review]
